Wrappers for lapack pstrf and pstf2, unit test included

tuxcell opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Add wrappers for lapack pstrf and pstf2
<!--Example: Closes gh-WXYZ.-->

#### To be used directly from scipy through linalg
<!--Please explain your changes.-->

#### To allow using pstrf and pstf2 (Cholesky factorization with complete pivoting of a positive semidefinite matrix A)
<!--Any additional information you think is important.-->

tuxcell commented:
import numpy as np
from scipy.linalg import cholesky

''' R code
> ( m <- matrix(c(5,1,1,3),2,2) )
     [,1] [,2]
[1,]    5    1
[2,]    1    3
> ( cm <- chol(m) )
         [,1]      [,2]
[1,] 2.236068 0.4472136
[2,] 0.000000 1.6733201
> ( cm <- chol(m, pivot = TRUE) )
         [,1]      [,2]
[1,] 2.236068 0.4472136
[2,] 0.000000 1.6733201
attr(,"pivot")
[1] 1 2
attr(,"rank")
[1] 2'''

m = np.array([[5,1],[1,3]])
c=cholesky(m)
print(c)

[[2.23606798 0.4472136 ]
 [0.         1.67332005]]

cholesky(m, chol_pivot=True)

(array([[2.23606798, 0.4472136 ],
        [0.        , 1.67332005]]), 2, array([1, 2], dtype=int32))

''' R code
> ( m <- matrix(c(6,  3,  4,  8,3,  6,  5,  1, 4,  5, 10,  7,8,  1,  7, 25),4,4) )
     [,1] [,2] [,3] [,4]
[1,]    6    3    4    8
[2,]    3    6    5    1
[3,]    4    5   10    7
[4,]    8    1    7   25
> ( cm <- chol(m) )
        [,1]     [,2]     [,3]      [,4]
[1,] 2.44949 1.224745 1.632993  3.265986
[2,] 0.00000 2.121320 1.414214 -1.414214
[3,] 0.00000 0.000000 2.309401  1.587713
[4,] 0.00000 0.000000 0.000000  3.132491
> ( cm <- chol(m, pivot = TRUE) )
     [,1]     [,2]     [,3]      [,4]
[1,]    5 1.400000 0.200000 1.6000000
[2,]    0 2.835489 1.664616 0.6207041
[3,]    0 0.000000 1.785792 0.9221487
[4,]    0 0.000000 0.000000 1.4847115
attr(,"pivot")
[1] 4 3 2 1
attr(,"rank")
[1] 4'''

m = np.array([[6,  3,  4,  8],[3,  6,  5,  1],[4,  5, 10,  7],[8,  1,  7, 25]])
cholesky(m)

array([[ 2.44948974,  1.22474487,  1.63299316,  3.26598632],
       [ 0.        ,  2.12132034,  1.41421356, -1.41421356],
       [ 0.        ,  0.        ,  2.30940108,  1.58771324],
       [ 0.        ,  0.        ,  0.        ,  3.13249102]])

cholesky(m, chol_pivot=True)

(array([[5.        , 1.4       , 0.2       , 1.6       ],
        [0.        , 2.83548938, 1.66461565, 0.62070414],
        [0.        , 0.        , 1.78579246, 0.92214868],
        [0.        , 0.        , 0.        , 1.48471148]]),
 4,
 array([4, 3, 2, 1], dtype=int32))

ilayn commented on scipy/linalg/decomp_cholesky.py:
```diff
@@ -41,10 +41,23 @@ def _cholesky(a, lower=False, overwrite_a=False, clean=True,
-    return c, lower
 
+    # if the pivot flag is false, return the result
+    if not chol_pivot:
```
This switch must happen earlier so you don't call both potrf and pstrf depending on the flag. You need to solve only once. Either one or the other.

tuxcell replied:
OK, fixed.

ilayn commented on scipy/linalg/flapack_pos_semidef.pyf.src:
```diff
@@ -0,0 +1,48 @@
+    ! P**T * A * P = L  * L**T,  if UPLO = 'L',
+    ! where U is an upper triangular matrix and L is lower triangular, and
+    ! P is stored as vector PIV.
+    callstatement (*f2py_func)((lower?"L":"U"),&n,a,&lda,piv,&rank_bn,&tol,work,&info); if(clean){int i,j;if(lower){for(i=0;i\<n;++i) for(j=i+1;j\<n;++j) *(a+j*n+i)=0.0;} else {for(i=0;i\<n;++i) for(j=i+1;j<n;++j) *(a+i*n+j)=0.0;}}
```
This is much easier on the python side to cleanup and creates a maintenance burden here.

tuxcell replied:
IMHO, cleaning code is best kept at the lowest level.

ilayn commented on scipy/linalg/flapack_pos_semidef.pyf.src:
```diff
@@ -0,0 +1,48 @@
+    integer optional,intent(in),check(clean==0||clean==1) :: clean = 1
+    integer depend(a),intent(hide):: n = shape(a,0)
+    <ftype2> dimension(n,n),check(shape(a,0)==shape(a,1)),intent(in,out,copy,out=c) :: a
+    integer, optional,check(shape(a,0)==lda),depend(a) :: lda=shape(a,0)
```
`LDA >= max(1,shape(a, 0))`

tuxcell replied:
OK, done.

ilayn commented on scipy/linalg/flapack_pos_semidef.pyf.src:
```diff
@@ -0,0 +1,48 @@
+! Signatures for f2py-wrappers of FORTRAN LAPACK Symmetric Positive Semidefinite Matrix functions.
+!
+
+subroutine <prefix2>pstrf(n,a,lda,piv,rank_bn,tol,work,info,lower,clean)
```
Why only prefix2? pstrf also has real and double flavors.

tuxcell replied:
OK, done.

ilayn commented on scipy/linalg/flapack_pos_semidef.pyf.src:
```diff
@@ -0,0 +1,48 @@
+    ! P**T * A * P = L  * L**T,  if UPLO = 'L',
+    ! where U is an upper triangular matrix and L is lower triangular, and
+    ! P is stored as vector PIV.
+    callstatement (*f2py_func)((lower?"L":"U"),&n,a,&lda,piv,&rank_bn,&tol,work,&info); if(clean){int i,j;if(lower){for(i=0;i\<n;++i) for(j=i+1;j\<n;++j) *(a+j*n+i)=0.0;} else {for(i=0;i\<n;++i) for(j=i+1;j<n;++j) *(a+i*n+j)=0.0;}}
```
This is much easier on the python side to cleanup and creates a maintenance burden here.

tuxcell replied:
IMHO, cleaning code is best kept at the lowest level.

ilayn commented on scipy/linalg/flapack_pos_semidef.pyf.src:
```diff
@@ -0,0 +1,48 @@
+    integer optional,intent(in),check(clean==0||clean==1) :: clean = 1
+    integer depend(a),intent(hide):: n = shape(a,0)
+    <ftype2> dimension(n,n),check(shape(a,0)==shape(a,1)),intent(in,out,copy,out=c) :: a
+    integer, optional,check(shape(a,0)==lda),depend(a) :: lda=shape(a,0)
```
`LDA >= max(1,shape(a, 0))`

tuxcell replied:
OK, done.

ilayn commented:
For the features you want to implement and to probe its applicability and scope, I would suggest opening an issue/asking it on the mailing list,  

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

tuxcell commented on scipy/linalg/flapack.pyf.src:
```diff
@@ -36,6 +36,7 @@ interface
     include 'flapack_gen_tri.pyf.src'
     include 'flapack_sym_herm.pyf.src'
     include 'flapack_pos_def.pyf.src'
+    include 'flapack_pos_semidef.pyf.src'
```
Hi @ilayn,
Thank you so much for your feedback.
You improved the the code, made it more clear and provided some good insight about the request process.
Your suggestion of checking for PSD is very relevant. I will add that now.
About leaving the method "cholesky" alone, may I suggest adding a method "cholesky_full_pivot" with this feature?

Shall I remove 'flapack_pos_semidef.pyf.src' and add its contents to 'flapack_pos_def.pyf.src'?

ilayn replied:
Yes that would be nice.

tuxcell commented on scipy/linalg/decomp_cholesky.py:
```diff
@@ -13,7 +13,7 @@
 
 def _cholesky(a, lower=False, overwrite_a=False, clean=True,
-              check_finite=True):
+              check_finite=True, chol_pivot=False, piv_tol=-1):
```
I renamed full_pivot and pivot_tol but the default value of -1 is the same used by R's "chol", I couldn't find a reason for that choice yet, maybe it's a very commonly used value?

tuxcell commented:
OK, thank you for the explanation.
I fixed the issues and added a few methods to test matrices for symmetry, PD, PSD, etc.
AFAIK, it looks fine now.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
         return S0, errest
+
+
+def is_matrix_symmetric(A, rtol=1e-05, atol=1e-08):
```
These helper functions won't fly in their current state. Especially with hard coded tolerances. Also hermitianness is needed here. Real versions are not enough.

tuxcell replied:
Added hermitianness and symmetry tested with a new method.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+
+    Examples
+    --------
+    from scipy.linalg import is_matrix_symmetric
```
every example code snippet needs to have `>>>` at the start of the line and then needs the results printed in the console. See other examples in the docs.

tuxcell replied:
OK, done

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    print(is_matrix_symmetric(m))
+    m = np.array([[5,0],[1,3]])
+    print(is_matrix_symmetric(m))
+    References
```
empty line needed.

tuxcell replied:
OK, done.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    print(is_matrix_symmetric(m))
+    References
+    ----------
+    Hazewinkel, Michiel, ed. (2001) [1994], "Symmetric matrix", Encyclopedia of Mathematics, 
```
I think symmetric matrices don't need any references.

tuxcell replied:
Removed references.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    https://stackoverflow.com/questions/42908334/checking-if-a-matrix-is-symmetric-in-numpy
+    """
+    A = _asarray_square(A)
+    return np.allclose(A, A.T, rtol=rtol, atol=atol)
```
This would fail with big and small entries arrays. It needs to be tied to the 1- and inf-norm of the array entries.

tuxcell replied:
Replaced with new method.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    if not is_matrix_symmetric(A):
+        raise ValueError('expected symmetric matrix')
+    eiVal, eiVecR=eig(A)
+    return (eiVal>0).all()
```
Same as above. If the array is too large it can always have some numerical noise around the small eigenvalues. Hence it is tricky for testing posdefness. That's why Cholesky decomposition is used for testing definiteness.

tuxcell replied:
Cholesky decomposition for testing definiteness has issues with round off errors that eigenvalues can overcome. Both solutions are far from ideal, however they are simple to implement. 
"In order to perform Cholesky Decomposition of a matrix, the matrix has to be a positive definite matrix." And to test for definiteness, one uses Cholesky Decomposition. It sounds recursive and it does not test for semi positive or for negative definiteness, which can be done with eigenvalues.

ilayn replied:
Semidefiniteness is much harder to test since it looks almost exact zeros.

ilayn replied:
Cholesky is cheaper to check but you are already eating away its advantage by computing eigenvalues which are more expensive.

tuxcell replied:
You're right. So, I changed the code accordingly.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    A = _asarray_square(A)
+    return np.allclose(A, A.T, rtol=rtol, atol=atol)
+
+def is_matrix_positive_definite(A):
```
Same as above.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+def is_matrix_symmetric(A, rtol=1e-05, atol=1e-08):
+    """
+    Returns True if a matrix is symmetric
+    Parameters
```
Empty lines.

tuxcell replied:
OK, done.

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    eiVal, eiVecR=eig(A)
+    return (eiVal>=0).all()
+
+def is_matrix_negative_definite(A):
```
Same as above

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,191 @@ def rounded_sign(x):
+    eiVal, eiVecR=eig(A)
+    return (eiVal<0).all()
+
+def is_matrix_negative_semidefinite(A):
```
This will fail for negative definite matrices

tuxcell replied:
Not anymore, now there is a tolerance argument.

tuxcell commented:
Thank you for the help with improving the code, it is much better now.

tuxcell commented:
import numpy as np

from scipy.linalg import is_matrix_hermitian,is_matrix_positive_definite, is_matrix_symmetric, is_matrix_positive_semidefinite,is_matrix_negative_definite,is_matrix_negative_semidefinite,is_matrix_indefinite

​

m = np.array([[5,1],[1,0]])

print(is_matrix_indefinite(m))

True

​

m = np.array([[-1,1],[1,-1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

m = np.array([[5,1],[1,3]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

True
True
False
False
False

m = np.array([[1, -1, 0],[-1, 5, 0],[0, 0, 7]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

True
True
False
False
False

m = np.array([[1,-1],[-1,1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
True
False
False
False

m = np.array([[-5,-1],[-1,-3]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
True
True
False

m = np.array([[-1,1],[1,-1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

m = np.array([[5,1],[1,0]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
False
True

m = np.array([[1,2],[2,1]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
False
True

m = np.array([[-2,1,-2],[1,-2,1],[-2,1,-2]])

print(is_matrix_positive_definite(m))

print(is_matrix_positive_semidefinite(m))

print(is_matrix_negative_definite(m))

print(is_matrix_negative_semidefinite(m))

print(is_matrix_indefinite(m))

False
False
False
True
False

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,236 @@ def rounded_sign(x):
+    return (eiVal<0).all()
+
+
+def is_matrix_negative_semidefinite(A, tol=1e-8):
```
`-np.eye(3)` returns true but shouldn't

ilayn commented on scipy/linalg/matfuncs.py:
```diff
@@ -668,3 +672,236 @@ def rounded_sign(x):
+    return (eiVal>0).all()
+
+
+def is_matrix_positive_semidefinite(A, tol=1e-8):
```
`np.eye(3)` returns true but shouldn't

tuxcell replied:
It's been removed

ilayn commented:
Hi @tuxcell . Thanks for the updates but you are trying to include a numerically unstable algorithm which is the detection of semidefiniteness. It is not possible to detect via `?pstrf` routines as you are trying to include now and to be honest I don't think it is possible at all without a rank revealing QR or SVD (which both are quite expensive). 

This is the same message that the authors of the routine gave [in their report](http://www.netlib.org/lapack/lawnspdf/lawn161.pdf) with a nice example of an indefinite matrix marked as positive semidefinite. 

Just to let you make aware of these results and remind that it would be much better if we only include the wrappers and avoid the helper functions.

tuxcell commented:
Hi @ilayn , thank you for the feedback. You are right, these are very expensive checks. I removed the helper functions, the changes in "cholesky" and left only the wrappers.
Thank you for  the report, it's really insightful.
Do I have to click on "close and comment" now?

tuxcell commented:
> I would strongly suggest that we do only the LAPACK wrappers in this PR and leave the rest to mailing list discussion or an extra PR. Because it introduces significant changes to cholesky and also new functions that are not necessarily valid to put in.

OK, I removed the helper functions, the changes in "cholesky" and left only the wrappers.

tuxcell commented:
Unit tests for both wrappers are included now.

ilayn commented:
Some pep8 issues causing hiccups

ilayn commented:
Hi @tuxcell I've tidied up a bit the wrappers and removed the cleaning parts since rank deficiency parts should also be zeroed out anyways and arguably better to do it by the user. Also that keeps the function signature compatible with native Fortran call. 

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches while keeping your master branch up to date with the official repo.

tuxcell commented:
Hi @ilayn, this is great, now I can start a few branches for a few ideas to explore.
Thank you for the very positive communication, it's been a quite interesting learning process.

ilayn commented:
My pleasure. Apparently the test I've included didn't go through. Let me quickly fix that and then we can hit the merge button.

tuxcell commented:
Allright! Have a nice weekend! :)

ilayn commented:
Ah this is why we need branches :) I'll revert the changes for a bit and then merge. Do you mind not adding changes to your master until we get it merged?

tuxcell commented:
OK, I'll be reading and playing with branches in the mean time and leave master alone.

ilayn commented:
OK all green and in it goes. Thanks @tuxcell !
